sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionManager.java

```diff
@@ -296,7 +302,10 @@ Mono<ServiceBusReceiveLink> getActiveLink() {
                     && ((AmqpException) failure).getErrorCondition() == AmqpErrorCondition.TIMEOUT_ERROR) {
                     return Mono.delay(SLEEP_DURATION_ON_ACCEPT_SESSION_EXCEPTION);
                 } else {
-                    return Mono.<Long>error(failure);
+                    // The link-endpoint-state publisher will emit signal on the reactor-executor thread, which is
+                    // non-blocking, if we use the session processor to recover the error, it requires a blocking
+                    // thread to close the client. Hence, we publish the error on the bounded elastic thread.
+                    return Mono.<Long>error(failure).publishOn(Schedulers.boundedElastic());
```

Comment:
Hi @anuchandy I suspect we haven't tested this error path before when using session processor. Because the link-endpoint-state publisher emits signal on reactor-executor thread, and processor recovery contains block() calls, so the the recovery would fail with error "block()/.. is not allowed on reactor-executor thread",
 
Currently, I think publish the error on the boundedElastic thread may involve less side effect, do you have any other ideals?

If we agree to add the publishOn(Schedulers.boundedElastic()) here,  we may also need to add for [the disposed error](https://github.com/Azure/azure-sdk-for-java/blob/dc387943322d149460803a0f2911fd872d8e45b7/sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/ServiceBusSessionManager.java#L296-L298).

Reply:
Hi Kun @liukun-msft, the apporach you coded looks good to me.

Given this is a rare & corner case, and we rely on the bounded thread pool for recovery, It would be helpful to have info logging to detect potential thread starvation by the pool. Something like, have the first log emission once we hit the error signal, then the second log emission once the error signal is picked by the bounded-elastic thread. If the receiver hangs with the first log entry, but without the second log entry from the bounded thread, that indicates thread starvation.

Hi Connie @conniey, I'm thinking about the offline chat we had on detecting thread starvation in the recovery route relying on the bounded pool; this is one of the routes. Tagging you if you have some thoughts.

Reply:
Sure, I'll add the info logs for tracking purpose. Do we have some standard log format for the tracking starvation issue? it can help us search and remove the logs once we resolve this problem in future. If no, we can take this as a start.

Reply:
Hmm..I think one stardard log format might not be achievable for all the cases. Maybe we can just include the keywords "bounded-elastic thread" for searching purpose. I simply added one log here, we can enhance it if we have more thoughts.

Reply:
We don't have any standard defined; as you mentioned, it's case by case. For this specific case, I was thinking something like (please review/consider if there is a better format)

```java
final String id = "trackingId:" + System.nanoTime();
logger.info("Unable to acquire new session. ({})", id, error);
Mono.<Long>error(failure).publishOn(Schedulers.boundedElastic()).doOnError(e -> {
   logger.info("Emitting the error signal received for session acquire attempt. {0}", id, error);
});
```

Reply:
And the logger should take care of printing the thread name/id

Reply:
Yes, add a tracking id in log is better if we need to handle multiple sessions 👍

Reply:
Right, in the case of more than one session error, if we see multiple entries of the first log and at least one of the second log entries, then it's good. 

It looks like one of the session errors resulting in the processor restart may cancel the merge operator, which could, in turn, cancel some  enqueued errors to bounded-pool that are yet-to-run; this is why I said at least one-second log entry (statement is based on code scanning, I didn't try to simulate it. Something to double check in future when we investigate the effect of handling amp-error in _RetryWhen_ level vs Processor level)